### PR TITLE
New feature: L4 Proxy support

### LINKED ISSUE #922: Support L4 proxy in edgemesh
**What would you like to be added**:
Add L4 proxy capability in the edgemesh 

**Why is this needed**:
In the current edgemesh implementation, REST is the only protocol supported. However, the native K8S service is L4 based. L4 service is necessary in many scenarios, like DB visiting.

### PR DESCRIPTION
**Pull request by arcanique**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
> New feature  support L4 proxy

**Which issue(s) this PR fixes**:

Fixes #922

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by arcanique** [comment]
cc/ @anyushun @qizha @sids-b

**Comment by sids-b** [comment]
@arcanique , apologies for delay in review... I will review it tomorrow

File: edgemesh/pkg/server/dns.go [hunk 1]
```diff
@@ -161,19 +165,26 @@ func recordHandler(que []dnsQuestion, req []byte) (rsp []byte, err error) {
...
 		}
 	}
+	fakeIP := defaultFakeIP
+	if ip != "" {
+		fakeIP = net.ParseIP(ip).To4()
+	}
 
-	//gen
-	pre, err := modifyRspPrefix(que)
+	pre, err := modifyRspPrefix(que, fakeIP)
```
Comment (anyushun) [inline edgemesh/pkg/server/dns.go]: Handle err please

File: edgemesh/pkg/server/dns.go [hunk 2]
```diff
@@ -161,19 +165,26 @@ func recordHandler(que []dnsQuestion, req []byte) (rsp []byte, err error) {
...
 	rsp = append(rsp, pre...)
 	for _, q := range que {
+		// head of each que is the same
+		if que[0].head.ansNum == 0 {
```
Comment (anyushun) [inline edgemesh/pkg/server/dns.go]: Why not ```q``` but ```que[0]```

File: edgemesh/pkg/proxy/proxy.go [hunk 3]
```diff
@@ -0,0 +1,564 @@
...
+	defaultNetworkPrefix     = "9.251."
+	defaultIpPoolSize        = 20
+	defaultTcpClientTimeout  = time.Second * 2
+	defaultTcpReconnectTimes = 3
+)
+
+var (
+	epoll         *poll.Epoll
+	unused        []string
+	serve         map[int32]listener
```
Comment (anyushun) [inline edgemesh/pkg/proxy/proxy.go]: Maybe ```sync.Map``` is better, just a suggestion

File: edgemesh/pkg/proxy/proxy.go [hunk 4]
```diff
@@ -0,0 +1,564 @@
...
+}
+
+type serviceTable struct {
+	ip         string
+	ports      []int32
+	targetPort []int32
+	lns        []net.Listener
+}
+
+type addrTable struct {
```
Comment (anyushun) [inline edgemesh/pkg/proxy/proxy.go]: Suggestion: ```sync.Map``` is better than ```map+RWMutex```

File: edgemesh/pkg/proxy/proxy.go [hunk 5]
```diff
@@ -0,0 +1,564 @@
...
+			break
+		}
+	}
+	// kubeedge edgemesh support bridge net ,so, use hostport to access
+	tp := st.targetPort[index]
+	targetPort := int32(0)
+	for _, value := range runPods[idx].Spec.Containers {
+		for _, v := range value.Ports {
+			if v.ContainerPort == tp {
+				targetPort = v.HostPort
```
Comment (rohitsardesai83) [inline edgemesh/pkg/proxy/proxy.go]: we need to break the loop in case of match.

Comment (arcanique) [reply edgemesh/pkg/proxy/proxy.go]: ok，thanks

File: edgemesh/pkg/proxy/virtualdevice/virtualdevice.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+		}
+		return err
+	}
+
+	return nil
+}
+
+// DestroyDevice implement release the device
+func DestroyDevice() error {
+	net.Interfaces()
```
Comment (rohitsardesai83) [inline edgemesh/pkg/proxy/virtualdevice/virtualdevice.go]: why is this call needed ?

Comment (arcanique) [reply edgemesh/pkg/proxy/virtualdevice/virtualdevice.go]: ok， I will remove this line

**Comment by sids-b** [comment]
@arcanique , can you please resolve the conflicts ?

**Comment by arcanique** [comment]
> @arcanique , can you please resolve the conflicts ?

ok

**Comment by arcanique** [comment]
> @arcanique , can you please resolve the conflicts ?

@sids-b Ok, solved

**Comment by sids-b** [comment]
Thanks @arcanique 

/approve

**Comment by sids-b** [comment]
@arcanique , can you please resolve the conflicts ?

File: edgemesh/pkg/proxy/proxy.go [hunk 7]
```diff
@@ -0,0 +1,570 @@
...
+	"fmt"
+	"math/rand"
+	"net"
+	"strconv"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/kubeedge/beehive/pkg/common/log"
```
Comment (subpathdev) [inline edgemesh/pkg/proxy/proxy.go]: could you please use klog instead?

File: edgemesh/pkg/server/dns.go [hunk 8]
```diff
@@ -15,21 +15,23 @@ import (
...
 )
 
 const (
-	aRecord   = 1
-	bufSize   = 1024
-	notImplem = uint16(0x0004)
+	aRecord       = 1
+	bufSize       = 1024
+	notImplem     = uint16(0x0004)
+	serverFAilure = uint16(0x0002)
```
Comment (subpathdev) [inline edgemesh/pkg/server/dns.go]: a small typo?
serverFAilure -> serverFailure

**Comment by sids-b** [comment]
ping @arcanique

**Comment by arcanique** [comment]
> ping @arcanique 
@sids-b Sorry. I will resolve the conflict as soon as possible tody

File: edgemesh/pkg/server/dns.go (hunk 8 above)
Comment (arcanique) [reply edgemesh/pkg/server/dns.go]: ok, thanks

File: edgemesh/pkg/proxy/proxy.go (hunk 7 above)
Comment (arcanique) [reply edgemesh/pkg/proxy/proxy.go]: sure

**Comment by sids-b** [comment]
Thanks @arcanique 

/lgtm

**Pull request merged**